app/feature/feature-payments/src/main/kotlin/com/hedvig/android/feature/payments/history/PaymentHistoryDestination.kt

```diff
@@ -35,47 +38,82 @@ import kotlinx.datetime.LocalDate
-          PaymentHistory.PastCharges.YearCharges(
-            year = year,
-            charge = charges.map { charge ->
-              PaymentHistory.PastCharges.YearCharges.Charge(
-                id = charge.id,
-                dueDate = charge.dueDate,
-                netAmount = charge.netAmount,
-                hasFailedCharge = charge.status == MemberCharge.MemberChargeStatus.FAILED,
-              )
-            },
-          )
-        },
-        showInfoAboutOlderCharges = pastCharges.size > 11,
-      ) to { chargeId: String ->
-        updatedOnChargeClicked(pastCharges.first { it.id == chargeId })
-      }
-    }
-  }
+  val uiState by viewModel.uiState.collectAsStateWithLifecycle()
   PaymentHistoryScreen(
-    paymentHistory = paymentHistory,
+    uiState = uiState,
     onChargeClicked = onChargeClicked,
     navigateUp = navigateUp,
+    reload = { viewModel.emit(PaymentHistoryEvent.Reload) },
   )
 }
 
 @Composable
 private fun PaymentHistoryScreen(
+  uiState: PaymentHistoryUiState,
+  onChargeClicked: (chargeId: String) -> Unit,
+  navigateUp: () -> Unit,
+  reload: () -> Unit,
+) {
+  when (uiState) {
+    PaymentHistoryUiState.Failure -> {
+      HedvigScaffold(
+        navigateUp = navigateUp,
+      ) {
+        HedvigErrorSection(onButtonClick = reload, modifier = Modifier.weight(1f))
+      }
+    }
+    PaymentHistoryUiState.Loading -> HedvigFullScreenCenterAlignedProgress()
+    is PaymentHistoryUiState.Success -> {
+      val updatedOnChargeClicked by rememberUpdatedState(onChargeClicked)
+      val (
+        paymentHistory: PaymentHistory, onChargeClickedAfterTransform: (
+          String,
+        ) -> Unit,
+      ) = remember(uiState.paymentHistory) {
+        if (uiState.paymentHistory.isEmpty()) {
+          PaymentHistory.NoHistoryData to { _ -> }
+        } else {
+          PaymentHistory.PastCharges(
+            chargesInYear = uiState.paymentHistory.sortedBy { it.dueDate }.groupBy { it.dueDate.year }.map {
+                (year, charges) ->
+              PaymentHistory.PastCharges.YearCharges(
+                year = year,
+                charge = charges.map { charge ->
+                  PaymentHistory.PastCharges.YearCharges.Charge(
+                    id = charge.id,
+                    dueDate = charge.dueDate,
+                    netAmount = charge.netAmount,
+                    hasFailedCharge = charge.status == MemberCharge.MemberChargeStatus.FAILED,
+                  )
+                },
+              )
+            },
+            showInfoAboutOlderCharges = uiState.paymentHistory.size > 11,
+          ) to { chargeId: String ->
+            updatedOnChargeClicked(uiState.paymentHistory.first { it.id == chargeId }.id)
+            // todo: I must say I'm a bit confused here :) but if understand correctly,
+            // we don't really need to update it anymore now that we have String as the parameter,
+            // we can just pass down onChargeClicked?
```

Comment:
we need to do `rememberUpdatedState` because it's inside a `remember` which does not itself include the lambda in its keys.
I think just leave it as it is, it should be fine :D

Reply:
Yeah in terms of correctness they should be giving the same end result

The one slight difference is that moving it as a key to the remember it will then trigger running the entire code inside the remember block, and it will produce the new values, and then potentially re-compose an extra time if those values are not stable etc etc. Nothing too important to be honest in this case here.